[Baskar Gopinath] iP

baskargopinath opened the pull request:
# Gops Chatbot
We will go through the following

* Chatbot commands
* Reason for choosing Gops as the name
* Design features

Before that, this is the steps we took to implement *Gops*
1. Fork the main ~~ip~~ Individual Project
2. Clone it into local
3. Make changes in IntelliJ 
4. And commit the changes

Guess the language we used
```java
public class Gops { ...
}
```

Which of these features would you like?
- [ ] Adding tasks to a list 
- [ ] Setting deadlines
- [ ] Clearing the list

> Guess what, **Gops** does all of them 😄 

Take a look at the [GitHub repository](https://github.com/bgopi23/ip) to see how it was implemented

Still don't know what language we used? Here's a hint **`Java`**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

baskargopinath committed:
Update gops.java

baskargopinath committed:
Update gops.java

baskargopinath committed:
Update gops.java

baskargopinath committed:
Update gops.java

baskargopinath committed:
Update gops.java

baskargopinath committed:
Automated Text UI Testing

baskargopinath committed:
Handle Errors

baskargopinath committed:
Update gops.java

baskargopinath committed:
Add support to save task list to hard disk

Added support to read task list from txt file

Added support to write task list to txt file

Includes feature to create new folder and file if data/gops.txt does not exist

baskargopinath committed:
Add support to take LocalDate as input

Deadline now takes a string of format yyyy-mm-dd and stores it as a actual LocalDate object and in the list as MMM-yyyy-dd

baskargopinath committed:
Reformat code with OOP

Reformatted code to extract out Ui, Storage, Parser and TaskList into each of their own classes

baskargopinath committed:
Move all classes into one package

Moved all classes into a packaged called Gops and refactored the code

baskargopinath committed:
Merge remote-tracking branch 'origin/add-gradle-support'

baskargopinath committed:
Add gradle support

Added gradle, deleted .idea file and rebuilt project

baskargopinath committed:
Add JUnit tests

Added tests to check if deadline and todo works properly

baskargopinath committed:
Create jar file using gradle

Created jar file using gradle by using shadowJar

baskargopinath committed:
Add JavaDoc comments

Added javadoc comments to 7/9 of the classes in Gops package

baskargopinath committed:
Tweak code based on java coding standard

Fixed errors raised by intellij java code formatter and adjusted code based on java code standard to the best of my knowledge

baskargopinath committed:
Add support to find tasks in list

Added support to find tasks with description that matches the search

kokerinks commented on src/main/java/Gops/gops.java:
Remember to make your class name and class file capitalised! should be Gops.java and 

```
public class gops {
```

----------------------------------------------------------
From further inspection, you might have gone with gops.java because it clashes with the package name. I think you can refactor and rename your **package name** to ```gops``` instead, as by coding standards package names should not be capitalised. That way, you will be able to capitalise your gops.java :)

kokerinks commented on src/main/java/Gops/Deadline.java:
```protected boolean todoStatus = false;```

1. Refer to this line from [coding standards](https://se-education.org/guides/conventions/java/intermediate.html):
> Boolean variables/methods should be named to sound like booleans

perhaps you can use a name like isDone, or isMarked?

2. Since the todoStatus is an attribute defined in Task already, it should be better practice to use it directly from the Task class,  instead of redefining it in Todo, Event and Deadline

kokerinks commented on src/main/java/Gops/Event.java:
Same as stated in Deadline:

> Since the todoStatus is an attribute defined in Task already, it should be better practice to use it directly from the Task class, instead of redefining it in Todo, Event and Deadline

kokerinks commented on src/main/java/Gops/Todo.java:
Same as stated in Deadline:

> Since the todoStatus is an attribute defined in Task already, it should be better practice to use it directly from the Task class, instead of redefining it in Todo, Event and Deadline

kokerinks reviewed (commented):
Aside from some minor code standard violations and deviation from OOP, everything else LGTM 👍

minreiseah commented on src/main/java/Gops/Parser.java:
```diff
@@ -0,0 +1,162 @@
+    public void parse(String input, TaskList taskList, File txtFile, Scanner inputTaker) {
+        String userReply = input;
+        while (userReply != null) {
+            if (userReply.equals("clear")) {
```
Perhaps abstract out your commands into a separate class/classes? For instance, you could have a `commandParser` class to parse just the command from the user class; and separate handlers for each command, as opposed to your logic being contained in this one `parse` method.

minreiseah commented on src/main/java/Gops/Parser.java:
```diff
@@ -0,0 +1,162 @@
+            } else if (userReply.contains("find")) {
+                String[] splitter = userReply.split(" ", 2);
+                System.out.println("Tasks that match your search");
+                System.out.println("------------------------------------------------------------------");
```
Maybe create a final object for your reusable print statements. Putting them into a separate Ui class would be great!

minreiseah commented on src/main/java/Gops/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+     * @param todoDescription task description
+     * @param status task status
+     */
+    public void addNewTodo(String todoDescription, boolean status) {
```
You might achieve better encapsulation if `TaskList` is agnostic to the types of tasks it can contain. So a general method of `addTask` may be more suitable. Though this is entirely up to your discretion.

minreiseah reviewed (approved):
Other than some light refactoring and naming, lgtm! :+1:

baskargopinath committed:
Change main class name

baskargopinath committed:
Add GUI to Gops Chatbot

added GUI and and completely changed parser so it doesn't print to cli and returns a string instead

baskargopinath committed:
Update README

The README.md was reformatted to enhance readability
and navigation.

Updated setup instructions are now easier to find

These changes make it easier to
setup Gops

baskargopinath committed:
Add assert statements

Added assert statements for enhanced code reliability and validation.

yisiox commented on src/main/java/Gops/Ui.java:
```diff
@@ -0,0 +1,44 @@
+            taskCount = taskList.numberOfTasks();
+        }
+        return taskList;
+        /*
+        System.out.println("Hello! I'm Gops");
+        System.out.println("What can I do for you?");
+        String userReply = inputTaker.nextLine();
+        parser.parse(userReply, taskList, txtFile, inputTaker);
+        if (userReply.equals("bye")) {
+            System.out.print("Bye. Hope to see you again soon!");
+        }
+         */
```
Should this be removed?

yisiox commented on src/main/java/Gops/TaskList.java:
```diff
@@ -0,0 +1,203 @@
+    public String listPrinter() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("Here are the tasks in your list\n");
+        stringBuilder.append("------------------------------------------------------------------\n");
```
Perhaps this can be replaced with e.g. `"-".repeat(50)` for readability. Could be it is stored as a constant since it is repeated elsewhere.

yisiox commented on src/main/java/Gops/Gops.java:
```diff
@@ -0,0 +1,142 @@
+import java.io.IOException;
+import java.util.Scanner;
+
+public class Gops extends Application {
```
Should public classes have JavaDoc header comments? I noticed the same issue in other files too.

yisiox commented on src/main/java/Gops/Gops.java:
```diff
@@ -0,0 +1,142 @@
+                ChatBox.getDukeDialog(dukeText, new ImageView(gops))
+        );
+        input.clear();
+    }
```
Nit: perhaps a line break between methods is good here.

yisiox commented on src/main/java/Gops/GopsException.java:
```diff
@@ -0,0 +1,8 @@
+    public GopsException() {
+        super();
+    }
+
```
Nit: unintended space.

yisiox commented on src/main/java/Gops/Parser.java:
```diff
@@ -0,0 +1,119 @@
+     * @param taskList list of tasks
+     * @param txtFile file containing tasks
+     */
+    public String parse(String userReply, TaskList taskList, File txtFile) {
```
This method is quite long, [issue 303](https://github.com/nus-cs2103-AY2324S2/forum/issues/303) on the forum may be  of interest.

yisiox commented on src/main/java/Gops/Storage.java:
```diff
@@ -0,0 +1,66 @@
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            if (taskWriter != null) {
```
Can a guard clause be used here?

yisiox commented on src/main/java/Gops/TaskList.java:
```diff
@@ -0,0 +1,203 @@
+        }
+        stringBuilder.append("------------------------------------------------------------------\n");
+        return stringBuilder.toString();
+    }
+    public void add(Todo task) {
+        taskList.add(task);
+    }
```
Nit: line breaks. I noticed the same issue in some other places too.

yisiox commented on src/main/java/Gops/Todo.java:
```diff
@@ -0,0 +1,23 @@
+        this.itemDescription = itemDescription;
+    }
+    public String doneOrNot() {
+        if (todoStatus) {
+            return "1";
+        } else {
+            return "0";
+        }
```
Nit: a ternary statement may be more concise.

yisiox commented on src/main/java/Gops/Todo.java:
```diff
@@ -0,0 +1,23 @@
+
+public class Todo {
+    protected String itemDescription;
+    protected boolean todoStatus = false;
```
Coding standard recommended to name boolean `isDone`. Noticed the same issue elsewhere.

yisiox commented on src/main/java/Gops/Storage.java:
```diff
@@ -0,0 +1,66 @@
+            taskReader = new BufferedReader(new FileReader(taskFile));
+            String taskString;
+            while ((taskString = taskReader.readLine()) != null) {
+                String[] taskData = taskString.split("\\|");
+                String taskType = taskData[0].trim();
+                boolean status = taskData[1].trim().equals("1");
+                String taskDescription = taskData[2].trim();
+
+                if (taskType.equals("T")) {
+                    taskList.addNewTodo(taskDescription, status);
+                } else if (taskType.equals("D")) {
+                    String endDate = taskData[3].trim().substring(3);
+                    taskList.addNewDeadline(taskDescription, endDate, status);
+                } else if (taskType.equals("E")) {
+                    String startBy = taskData[3].trim().substring(6);
+                    String endBy = taskData[4].trim().substring(4);
+                    taskList.addNewEvent(taskDescription, startBy, endBy, status);
+                }
```
Perhaps this part can be SLAP'd.

yisiox commented on src/main/java/Gops/Event.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    @Override
+    public String stringPrinter() {
+        return "E " + "| " + this.doneOrNot() + " | " + itemDescription + " | from: " + startBy + " | to: " + endBy + "";
```
Is this line over 120 characters long?

yisiox reviewed (commented):
Did my best to add code quality related comments, but could have missed some. LGTM overall I was able to follow quite well. Hope I was able to help!

baskargopinath committed:
Improve code quality

Refactored todoStatus to isDone.

Removed isDone from Deadline
and Event.

Added welcome message

baskargopinath committed:
Merge pull request #2 from bgopi23/branch-A-CodeQuality

Improve code quality

baskargopinath committed:
Add help extension

Added help extension
to guide user

baskargopinath committed:
Update UG

Updated user guide

baskargopinath committed:
Update build-gradle

Updated build-gradle

baskargopinath committed:
Add launcher

Added launcher class
